## ALGUNOS EJEMPLOS DE ALGEBRA MATRICIAL EN JULIA
### MECÁNICA DE SÓLIDOS
### Ramón Zúñiga, CGEO UNAM

In [112]:
using LinearAlgebra

Supongamos una matriz

\begin{equation}\boldsymbol A =
\begin{bmatrix} 4 & -2  \\
                1 & 1  \\
\end{bmatrix}\end{equation}


In [113]:
A = [
    4 -2
    1 1
]

2×2 Matrix{Int64}:
 4  -2
 1   1

In [114]:
inv(A) # Inversa de la Matriz A

2×2 Matrix{Float64}:
  0.166667  0.333333
 -0.166667  0.666667

Confirmamos que

$ A \cdot A^{-1} = I $

In [115]:
A*inv(A)

2×2 Matrix{Float64}:
 1.0          0.0
 2.77556e-17  1.0

In [116]:
A' # Adjunta de la matriz A

2×2 adjoint(::Matrix{Int64}) with eltype Int64:
  4  1
 -2  1

In [117]:
det(A) # Determinante de A

6.0

In [118]:
transpose(A)  # Transpuesta de A

2×2 transpose(::Matrix{Int64}) with eltype Int64:
  4  1
 -2  1

### Valores Propios y Vectores Propios (Eigenvalores y Eigenvectores)

Los valores propios λ y los vectores propios v
  son las soluciones de
    
\begin{equation}
(\boldsymbol{\sigma}-\mathbf{I}\lambda)\mathbf{v}
\tag{5} = 0
\end{equation}

In [119]:
λ = eigvals(A)

2-element Vector{Float64}:
 2.0
 3.0

In [120]:
v = eigvecs(A)

2×2 Matrix{Float64}:
 0.707107  0.894427
 0.707107  0.447214

Lo que sigue es para confirmar la operación fundamental de la matrix con un eigenvector que debe dar:

\begin{equation}
\boldsymbol{\sigma}\mathbf{v}=\lambda\mathbf{v}
\end{equation}

In [121]:
A*v[:,1]  # Esto multiplica la matriz A por el primer eigenvector

2-element Vector{Float64}:
 1.414213562373095
 1.414213562373095

Notar que los dos valores anteriores son únicamente los valores del eigenvector por el primer eigenvalor
0.707107*2.0  # los dos valores del eigenvector 1 son iguales

In [123]:
A*v[:,2]  # Ahora multiplicamos por el segundo eigenvector

2-element Vector{Float64}:
 2.6832815729997477
 1.3416407864998738

In [124]:
# en el caso del segundo eigenvector tenemos dos valores diferentes, pero el resultado es igual a multiplicarlos
# por el segundo eigenvalor

In [125]:
0.894427*3

2.683281

In [126]:
0.447214*3

1.341642

In [127]:
B = rand(2,1)

2×1 Matrix{Float64}:
 0.8667962538562278
 0.44496969848662116

In [128]:
C = A*B

2×1 Matrix{Float64}:
 2.577245618451669
 1.3117659523428489

In [129]:
D = B*A # aquí confirmamos que la multiplicación no es conmutativa

LoadError: DimensionMismatch: matrix A has dimensions (2,1), matrix B has dimensions (2,2)

In [130]:
E = rand(2,2)

2×2 Matrix{Float64}:
 0.713446   0.518419
 0.0931893  0.80004

In [131]:
F = A*E

2×2 Matrix{Float64}:
 2.66741   0.473594
 0.806635  1.31846

In [132]:
G = E*A

2×2 Matrix{Float64}:
 3.3722  -0.908473
 1.1728   0.613661

#### Escritura y Lectura de Datos Tabulares
#### Creamos datos (matriz A), los guardamos en un archivo con el nombre de DatosA especificando espaciador con blancos, y después lo volvemos a leer como la matriz B

In [133]:
using DelimitedFiles
A = rand(50,2) # Creamos datos al azar, por ejemplo, una matriz de 50 filas y dos columnas

writedlm("DatosA", A, ' ')  # Escribimos los datos con espacios en un archivo de nombre DatosA 
B = readdlm("DatosA", ' ');   # Leemos los datos del archivo indicando que los espacios son separadores
